src/mapping/query/function/contains.rs

```diff
@@ -0,0 +1,168 @@
+use super::prelude::*;
+
+#[derive(Debug)]
+pub(in crate::mapping) struct ContainsFn {
+    query: Box<dyn Function>,
+    substring: Box<dyn Function>,
+    case_sensitive: Option<Box<dyn Function>>,
+}
+
+impl ContainsFn {
+    #[cfg(test)]
+    pub(in crate::mapping) fn new(
+        query: Box<dyn Function>,
+        substring: &str,
+        case_sensitive: bool,
+    ) -> Self {
+        let substring = Box::new(Literal::from(Value::from(substring)));
+        let case_sensitive = Some(Box::new(Literal::from(Value::from(case_sensitive))) as _);
+
+        Self {
+            query,
+            substring,
+            case_sensitive,
+        }
+    }
+}
+
+impl Function for ContainsFn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        let substring = match self.substring.execute(ctx)? {
+            Value::Bytes(b) => b,
+            v => unexpected_type!(v),
+        };
+
+        let value = match self.query.execute(ctx)? {
+            Value::Bytes(b) => b,
+            v => unexpected_type!(v),
+        };
+
+        let mut contains = value
+            .windows(substring.len())
+            .any(|window| window == substring);
```

Comment:
Note that this _may be_ slower than converting with `from_utf8_lossy` and using `contains`, as the string matching could use a faster algorithm (ie Boyer-Moore search, etc) than comparing the string at each location. Which one would end up being faster, I can't know without testing. If you go that route, it may be best to handle the case insensitive match in one step instead of testing case sensitive first and then converting.